File: Cpp/160_Intersection_of_Two_Linked_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int headA_length = getLength(headA);
        int headB_length = getLength(headB);
        
        ListNode* nodeA = headA;
        ListNode* nodeB = headB;
        int diff = headA_length - headB_length;
        if(headA_length < headB_length) {
            nodeA = headB;
            nodeB = headA;
            diff = headB_length - headA_length;
        }
        
        for(int i=0; i<diff; ++i){
          nodeA = nodeA->next;
        }
        
        while(nodeA != nodeB){
            if(!nodeA || !nodeB) return nullptr;
            nodeA = nodeA->next;
            nodeB = nodeB->next;
        }
        
        return nodeA;
    }
    
    int getLength(ListNode* head){
        ListNode* node = head;
        int length = 0;
        while(node){
            ++length;
            node = node->next;
        }
        
        return length;
    }
};

File: Kotlin/160_Intersection_of_Two_Linked_Lists.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun getIntersectionNode(headA:ListNode?, headB:ListNode?):ListNode? {
        var nodeA = headA
        var nodeB = headB
        
        while (nodeA != nodeB) {
            nodeA = if (nodeA == null) headB else nodeA.next
            nodeB = if (nodeB == null) headA else nodeB.next
        }
        
        return nodeA
    }
}